crates/bevy_ecs/src/world/archetype_invariants.rs

```diff
@@ -0,0 +1,337 @@
+    NoneOf(PhantomData<B>),
+}
+
+impl<B: Bundle> ArchetypeStatement<B> {
+    /// Erases the type information of this archetype statment.
+    ///
+    /// Requires mutable world access, since the components might not have been added to the world yet.
+    pub fn into_untyped(self, world: &mut World) -> UntypedArchetypeStatement {
+        let component_ids = B::component_ids(&mut world.components, &mut world.storages);
+        let component_ids: HashSet<ComponentId> = component_ids.into_iter().collect();
+
+        match self {
+            ArchetypeStatement::AllOf(_) => UntypedArchetypeStatement::AllOf(component_ids),
+            ArchetypeStatement::AtLeastOneOf(_) => {
+                if component_ids.len() == 1 {
+                    warn!("An `ArchetypeStatement::AtLeastOneOf` was constructed for a bundle with only one component. Prefer the equivalent `ArchetypeStatment:AllOf` for consistency and clarity.");
+                }
+                UntypedArchetypeStatement::AtLeastOneOf(component_ids)
+            }
+            ArchetypeStatement::AtMostOneOf(_) => {
+                UntypedArchetypeStatement::AtMostOneOf(component_ids)
+            }
+            ArchetypeStatement::NoneOf(_) => UntypedArchetypeStatement::NoneOf(component_ids),
+        }
+    }
+
+    /// Constructs a new [`ArchetypeStatement::AllOf`] variant for all components stored in the bundle `B`
+    #[inline]
+    pub const fn all_of() -> Self {
+        ArchetypeStatement::AllOf(PhantomData)
+    }
+
+    /// Constructs a new [`ArchetypeStatement::AtLeastOneOf`] variant for all components stored in the bundle `B`
+    #[inline]
+    pub const fn at_least_one_of() -> Self {
+        ArchetypeStatement::AtLeastOneOf(PhantomData)
+    }
+
+    /// Constructs a new [`ArchetypeStatement::AtMostOneOf`] variant for all components stored in the bundle `B`
+    #[inline]
+    pub const fn at_most_one_of() -> Self {
+        ArchetypeStatement::AtMostOneOf(PhantomData)
+    }
+
+    /// Constructs a new [`ArchetypeStatement::NoneOf`] variant for all components stored in the bundle `B`
+    #[inline]
+    pub const fn none_of() -> Self {
+        ArchetypeStatement::NoneOf(PhantomData)
+    }
+}
+
+/// A type-erased version of [`ArchetypeInvariant`].
+///
+/// Intended to be used with dynamic components that cannot be represented with Rust types.
+/// Prefer [`ArchetypeInvariant`] when possible.
+#[derive(Clone, Debug, PartialEq)]
+pub struct UntypedArchetypeInvariant {
+    /// For all entities where the predicate is true
+    pub predicate: UntypedArchetypeStatement,
+    /// The consequence must also be true
+    pub consequence: UntypedArchetypeStatement,
+}
+
+impl UntypedArchetypeInvariant {
+    /// Assert that the provided iterator of [`ComponentId`]s obeys this archetype invariant
+    ///
+    /// `component_ids` is generally provided via the `components` field on [`Archetype`](crate::archetype::Archetype).
```

Comment:
I think it's useful to include a signpost here, to point users in the correct direction. Also, the component ids for an archetype are fetched by a method on `Archetype`, not `ComponentId`

Reply:
My concern is that the position of the statement is very vulnerable to updates. If the best way to obtain `ComponentId` is ever going to change (and it happened, with #5066) this piece of information goes out of date and it will be difficult that anybody will realize that.

The best way is to keep the discussion relevant to the item, and use links instead for other pieces of information (this is already done by the link in the first line).

Reply:
Hmm. Perhaps we can add a pointer to the method on `Bundle` from the `ComponentId` docs?

Reply:
Do you mean the method on `Archetype`? I think that's possibly the right way to go, but ideally the `Archetype` docs themselves would make this clear. After thinking on it, I think the right move here is to remove the reference in this PR, and punt the explanation to the docs for `Archetype` once they get written. Does that sound good to y'all?

Reply:
Yeah, SGTM.

Reply:
Removed as of 835f7cdbbd1128d8c208ca60a003de222b8983ac